Bottom: d687c2a7b298467031ac36a0a51054692cfcabad
Top:    5564f68fab90e2c0425bdc83716f4c386fe834b8
Author: Ineiev <ineiev@users.sourceforge.net>
Date:   2011-01-14 18:22:49 +0000

hid/gtk: Round off digits in coordinate display

Fix annoying coordinates such as this:
Example (grid set to 1mm): 116.9998 43.9999

Closes-bug: lp-699169


---

diff --git a/src/hid/gtk/gui-misc.c b/src/hid/gtk/gui-misc.c
index e3316aa..7e71268 100644
--- a/src/hid/gtk/gui-misc.c
+++ b/src/hid/gtk/gui-misc.c
@@ -548,38 +548,45 @@ ghid_set_status_line_label (void)
 }
 
 /* ---------------------------------------------------------------------------
+ * transforms a pcb coordinate to selected units
+ * adjusted to the nearest grid point.
+ */
+static double
+grid_pcb_to_units (double x)
+{
+  if (Settings.grid_units_mm)
+  return floor (x / PCB->Grid + .5) * PCB->Grid * COOR_TO_MM;
+  return x / 100;/* adjustment is not needed
+    probably because x / 100 is always 'integer' enough. */
+}
+
+/* ---------------------------------------------------------------------------
  * output of cursor position
  */
 void
 ghid_set_cursor_position_labels (void)
 {
   gchar text[128];
+  int prec = Settings.grid_units_mm ? 4: 2;
 
   if (Marked.status)
-    {double scale, dx, dy, r, a;
-      scale = Settings.grid_units_mm ? COOR_TO_MM: 1. / 100;
-      dx = (Crosshair.X - Marked.X) * scale;
-      dy = (Marked.Y - Crosshair.Y) * scale;
-      r = sqrt( dx * dx + dy * dy);
-      a = atan2(dy, dx) * 180 / M_PI;
-      if (Settings.grid_units_mm)
-	snprintf (text, sizeof (text), "r %-.4f; phi %-.1f; %-.4f %-.4f",
-		  r, a, dx, dy);
-
-      else
-	snprintf (text, sizeof (text), "r %-.2f; phi %-.1f; %-.2f %-.2f",
-		  r, a, dx, dy);
+    {
+      double dx, dy, r, a;
+
+      dx = grid_pcb_to_units (Crosshair.X - Marked.X);
+      dy = grid_pcb_to_units (Crosshair.Y - Marked.Y);
+      r = sqrt (dx * dx + dy * dy);
+      a = atan2 (dy, dx) * 180 / M_PI;
+      snprintf (text, sizeof (text), "r %-.*f; phi %-.1f; %-.*f %-.*f",
+		prec, r, a, prec, dx, prec, dy);
       ghid_cursor_position_relative_label_set_text (text);
     }
   else
     ghid_cursor_position_relative_label_set_text ("r __.__; phi __._; __.__ __.__");
 
-  if (Settings.grid_units_mm)
-    snprintf (text, sizeof (text), "%-.4f %-.4f",
-	      COOR_TO_MM * Crosshair.X, COOR_TO_MM * Crosshair.Y);
-  else
-    snprintf (text, sizeof (text), "%-.2f %-.2f",
-	      Crosshair.X / 100.,  Crosshair.Y / 100.);
+  snprintf (text, sizeof (text), "%-.*f %-.*f",
+            prec, grid_pcb_to_units (Crosshair.X),
+            prec, grid_pcb_to_units (Crosshair.Y));
 
   ghid_cursor_position_label_set_text (text);
 }
